File: ai/regression_plot.py
#!/usr/bin/env python3
"""Linear regression analysis of salary versus years of experience."""

import os
from pathlib import Path

cache_dir = Path(__file__).parent / ".matplotlib_cache"
cache_dir.mkdir(exist_ok=True)
os.environ.setdefault("MPLCONFIGDIR", str(cache_dir))

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

SCRIPT_DIR = Path(__file__).parent
DATA_FILE = SCRIPT_DIR / "regression_data.csv"
OUTPUT_FILE = SCRIPT_DIR / "linear_regression_python_output.png"

X_COLUMN = "YearsExperience"
Y_COLUMN = "Salary"
PREDICT_YEARS = 6.0


def format_currency(value: float) -> str:
    return f"${value:,.0f}"


def create_regression_plot(
    data: pd.DataFrame,
    model: LinearRegression,
    intercept: float,
    slope: float,
    r2: float,
    predict_years: float,
    predicted_salary: float,
    output_file: Path,
) -> None:
    x_values = data[X_COLUMN].values
    y_values = data[Y_COLUMN].values
    x_line = np.linspace(x_values.min(), x_values.max(), 100).reshape(-1, 1)
    y_line = model.predict(x_line)

    fig, ax = plt.subplots(figsize=(12, 6))
    fig.patch.set_facecolor("#fafafa")
    ax.set_facecolor("#ffffff")

    ax.scatter(
        x_values,
        y_values,
        s=90,
        alpha=0.85,
        color="#2E86AB",
        edgecolors="white",
        linewidths=1.2,
        label="Observed data",
        zorder=3,
    )
    ax.plot(
        x_line,
        y_line,
        color="#C0392B",
        linewidth=2.5,
        label="Fitted regression line",
        zorder=2,
    )
    ax.scatter(
        [predict_years],
        [predicted_salary],
        s=180,
        color="#F39C12",
        edgecolors="#2C3E50",
        linewidths=1.5,
        label=f"Prediction at {predict_years:g} years",
        zorder=4,
    )

    ax.annotate(
        f"Predicted salary: {format_currency(predicted_salary)}",
        xy=(predict_years, predicted_salary),
        xytext=(predict_years + 0.35, predicted_salary + 3500),
        fontsize=11,
        fontweight="bold",
        color="#2C3E50",
        arrowprops={"arrowstyle": "->", "color": "#2C3E50", "lw": 1.2},
        bbox={"boxstyle": "round,pad=0.4", "facecolor": "#FFF3CD", "edgecolor": "#F39C12"},
    )

    equation = f"Salary = {intercept:,.2f} + {slope:,.2f} × YearsExperience"
    stats_text = f"{equation}\nR² = {r2:.4f}"
    ax.text(
        0.02,
        0.98,
        stats_text,
        transform=ax.transAxes,
        va="top",
        ha="left",
        fontsize=11,
        bbox={"boxstyle": "round,pad=0.5", "facecolor": "white", "edgecolor": "#cccccc", "alpha": 0.95},
    )

    ax.set_title(
        "Linear Regression: Salary vs Years of Experience",
        fontsize=15,
        fontweight="bold",
        pad=14,
    )
    ax.set_xlabel("Years of Experience", fontsize=12)
    ax.set_ylabel("Salary ($)", fontsize=12)
    ax.grid(True, linestyle="--", alpha=0.35)
    ax.legend(loc="lower right", frameon=True, framealpha=0.95)
    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)

    fig.tight_layout()
    fig.savefig(output_file, dpi=150, bbox_inches="tight")
    plt.close(fig)


def main() -> None:
    if not DATA_FILE.exists():
        print(f"Error: file not found: {DATA_FILE}")
        raise SystemExit(1)

    data = pd.read_csv(DATA_FILE)
    if X_COLUMN not in data.columns or Y_COLUMN not in data.columns:
        print(f"Error: required columns not found in {DATA_FILE.name}")
        print(f"Available columns: {', '.join(data.columns)}")
        raise SystemExit(1)

    x = data[[X_COLUMN]].values
    y = data[Y_COLUMN].values

    model = LinearRegression()
    model.fit(x, y)
    y_pred = model.predict(x)

    intercept = float(model.intercept_)
    slope = float(model.coef_[0])
    r2 = float(r2_score(y, y_pred))
    predicted_salary = float(model.predict([[PREDICT_YEARS]])[0])

    equation = f"Salary = {intercept:.4f} + {slope:.4f} * YearsExperience"

    print("Linear Regression Results")
    print("-" * 40)
    print(f"Slope: {slope:.4f}")
    print(f"Intercept: {intercept:.4f}")
    print(f"Regression equation: {equation}")
    print(f"R-squared: {r2:.4f}")
    print(
        f"Predicted salary at {PREDICT_YEARS:g} years of experience: "
        f"{predicted_salary:.2f}"
    )

    create_regression_plot(
        data=data,
        model=model,
        intercept=intercept,
        slope=slope,
        r2=r2,
        predict_years=PREDICT_YEARS,
        predicted_salary=predicted_salary,
        output_file=OUTPUT_FILE,
    )

    print(f"Saved regression plot to {OUTPUT_FILE.name}")


if __name__ == "__main__":
    main()

File: ai/regression_plot.R
#!/usr/bin/env Rscript
# Linear regression analysis of salary versus years of experience.

script_dir <- if (sys.nframe() == 0) {
  dirname(normalizePath(sub("--file=", "", commandArgs(trailingOnly = FALSE)[grep("--file=", commandArgs(trailingOnly = FALSE))][1])))
} else {
  "."
}

data_file <- file.path(script_dir, "regression_data.csv")
output_file <- file.path(script_dir, "linear_regression_r_output.png")

x_column <- "YearsExperience"
y_column <- "Salary"
predict_years <- 6

format_currency <- function(value) {
  paste0("$", format(round(value), big.mark = ",", scientific = FALSE, trim = TRUE))
}

if (!file.exists(data_file)) {
  cat(sprintf("Error: file not found: %s\n", data_file))
  quit(status = 1)
}

data <- read.csv(data_file, stringsAsFactors = FALSE)

if (!(x_column %in% names(data)) || !(y_column %in% names(data))) {
  cat(sprintf("Error: required columns not found in %s\n", basename(data_file)))
  cat(sprintf("Available columns: %s\n", paste(names(data), collapse = ", ")))
  quit(status = 1)
}

formula <- as.formula(paste(y_column, "~", x_column))
model <- lm(formula, data = data)

coefs <- coef(model)
intercept <- unname(coefs[1])
slope <- unname(coefs[2])
r2 <- summary(model)$r.squared
predicted_salary <- as.numeric(predict(model, newdata = data.frame(YearsExperience = predict_years)))
equation <- sprintf("Salary = %.4f + %.4f * YearsExperience", intercept, slope)

cat("Linear Regression Results\n")
cat(paste(rep("-", 40), collapse = ""), "\n", sep = "")
cat(sprintf("Slope: %.4f\n", slope))
cat(sprintf("Intercept: %.4f\n", intercept))
cat(sprintf("Regression equation: %s\n", equation))
cat(sprintf("R-squared: %.4f\n", r2))
cat(sprintf(
  "Predicted salary at %g years of experience: %.2f\n",
  predict_years,
  predicted_salary
))

x_vals <- data[[x_column]]
y_vals <- data[[y_column]]
x_line <- seq(min(x_vals), max(x_vals), length.out = 100)
y_line <- intercept + slope * x_line

png(output_file, width = 1200, height = 600, res = 150)
par(
  mar = c(5, 5, 4, 2),
  bg = "#FAFAFA",
  fg = "#2C3E50",
  col.axis = "#2C3E50",
  col.lab = "#2C3E50",
  col.main = "#2C3E50",
  font.lab = 2,
  font.main = 2,
  las = 1
)

plot(
  x_vals,
  y_vals,
  pch = 19,
  cex = 1.35,
  col = adjustcolor("#2E86AB", alpha.f = 0.85),
  xlab = "Years of Experience",
  ylab = "Salary ($)",
  main = "Linear Regression: Salary vs Years of Experience",
  panel.first = grid(col = "#D5D8DC", lty = "dashed")
)

lines(x_line, y_line, col = "#C0392B", lwd = 2.5)
points(
  predict_years,
  predicted_salary,
  pch = 21,
  bg = "#F39C12",
  col = "#2C3E50",
  cex = 2.2,
  lwd = 1.5
)

text(
  predict_years + 0.35,
  predicted_salary + 3500,
  labels = paste0("Predicted salary: ", format_currency(predicted_salary)),
  pos = 4,
  font = 2,
  cex = 0.95,
  col = "#2C3E50"
)

stats_label <- paste0(
  sprintf("Salary = %s + %s × YearsExperience", format(intercept, big.mark = ",", nsmall = 2), format(slope, big.mark = ",", nsmall = 2)),
  sprintf("\nR² = %.4f", r2)
)
legend(
  "bottomright",
  legend = c("Observed data", "Fitted regression line", sprintf("Prediction at %g years", predict_years)),
  col = c("#2E86AB", "#C0392B", "#F39C12"),
  pch = c(19, NA, 21),
  lty = c(NA, 1, NA),
  lwd = c(NA, 2.5, NA),
  pt.bg = c(NA, NA, "#F39C12"),
  pt.cex = c(1.2, NA, 1.6),
  bty = "n",
  cex = 0.95
)
rect(
  grconvertX(0.02, "npc"),
  grconvertY(0.98, "npc"),
  grconvertX(0.46, "npc"),
  grconvertY(0.84, "npc"),
  col = adjustcolor("white", alpha.f = 0.95),
  border = "#CCCCCC"
)
text(
  grconvertX(0.03, "npc"),
  grconvertY(0.965, "npc"),
  labels = stats_label,
  adj = c(0, 1),
  cex = 0.95
)

dev.off()

cat(sprintf("Saved regression plot to %s\n", basename(output_file)))
